src/crimson/os/seastore/cache.cc

```diff
@@ -0,0 +1,220 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "crimson/os/seastore/cache.h"
+#include "crimson/common/log.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore {
+
+Cache::~Cache()
+{
+  retire_extent(root);
+  root.reset();
+  for (auto i = dirty.begin(); i != dirty.end(); ) {
+    auto ptr = &*i;
+    dirty.erase(i++);
+    intrusive_ptr_release(ptr);
+  }
+  for (auto &i: extents) {
+    logger().error("~Cache: extent {} still alive", i);
+  }
+  ceph_assert(extents.empty());
+}
+
+void Cache::add_extent(CachedExtentRef ref)
+{
+  assert(ref->is_valid());
+  extents.insert(*ref);
+
+  ceph_assert(!ref->primary_ref_list_hook.is_linked());
+  if (ref->is_dirty()) {
+    intrusive_ptr_add_ref(&*ref);
+    dirty.push_back(*ref);
+  }
+  logger().debug("add_extent: {}", *ref);
+}
+
+void Cache::retire_extent(CachedExtentRef ref)
+{
+  logger().debug("retire_extent: {}", *ref);
+  assert(ref->is_valid());
+  extents.erase(*ref);
+
+  if (ref->is_dirty()) {
+    ceph_assert(ref->primary_ref_list_hook.is_linked());
+    dirty.erase(dirty.s_iterator_to(*ref));
+    intrusive_ptr_release(&*ref);
+  } else {
+    ceph_assert(!ref->primary_ref_list_hook.is_linked());
+  }
+}
+
+CachedExtentRef Cache::duplicate_for_write(
+  Transaction &t,
+  CachedExtentRef i) {
+  if (i->is_pending())
+    return i;
+
+  auto ret = i->duplicate_for_write();
+  ret->version++;
+  ret->state = CachedExtent::extent_state_t::MUTATION_PENDING;
+
+  if (ret->get_type() == extent_types_t::ROOT) {
+    t.root = ret->cast<RootBlock>();
+  }
+
+  t.add_to_retired_set(i);
+  t.add_mutated_extent(ret);
+
+  return ret;
+}
+
+std::optional<record_t> Cache::try_construct_record(Transaction &t)
+{
+  // First, validate read set
+  for (auto &i: t.read_set) {
+    if (i->state == CachedExtent::extent_state_t::INVALID)
+      return std::nullopt;
+  }
+
+  record_t record;
+
+  // Transaction is now a go, set up in-memory cache state
+  // invalidate now invalid blocks
+  for (auto &i: t.retired_set) {
+    logger().debug("try_construct_record: retiring {}", *i);
+    ceph_assert(!i->is_pending());
+    ceph_assert(i->is_valid());
+    retire_extent(i);
+    i->state = CachedExtent::extent_state_t::INVALID;
+  }
+
+  t.write_set.clear();
+
+  // Add new copy of mutated blocks, set_io_wait to block until written
+  record.deltas.reserve(t.mutated_block_list.size());
+  for (auto &i: t.mutated_block_list) {
+    logger().debug("try_construct_record: mutating {}", *i);
+    add_extent(i);
+    i->prepare_write();
+    i->set_io_wait();
+    record.deltas.push_back(
+      delta_info_t{
+	i->get_type(),
+	i->get_paddr(),
+	(segment_off_t)i->get_length(),
+	i->get_version(),
+	i->get_delta()
```

Comment:
@athanatos i just want to make sure that by collecting deltas in this way does not necessarily imply that

> if mutated block 42 does not have a delta, we can assume that the delta in this transaction does not change it as a sequence of `apply_delta(delta_from_mutated_block_34)` where block 34 is mutated in the same transaction.

right? as i think we can have a just-enough set of deltas which only record the operations instead of their consequences. in the context of onode tree changes, there could be just 2 types of deltas:

1. assign
1. remove

 there could be more finer grained deltas in future, like mutating the embedded extent map, or xattr map. but i think we can just attach these deltas to the onode block even if applying them could cause mutations to other blocks addressed with their logical addresses.

Reply:
No, if block 42 (physical) changes, there must be a delta for block 42.  The idea here is that cache doesn't really understand lbas at all, the layer above is responsible for ensuring that any mutated extent carries a delta describing the mutation.

At a higher level, consider assignment.  If the original transaction linked a particular onode into the onode tree and it caused a split, there are corresponding updates to both the lba tree and other onode tree internals.  Replaying a logical insertion certainly wouldn't cause the new nodes to be allocated at the same physical addresses, so the lba pointers would be wrong in a way that would be hard to fix in a general fashion.

As such, the idea here is purely that for any cached, dirty block, replaying journal entries at when the osd is restarted will result in the same block with the same contents without the cache needing to know anything about higher level structures.

As a side effect, tracking mutations at this granularity is also how we detect transaction conflicts of the sort mentioned by Yingxin in his email.

Reply:
i see. so a possible `apply_delta()` for onode block could be something like:

```c++
void apply_delta(paddr_t base, ceph::bufferlist &bl) final
{
  // cast the its internal node_t to pointer returned by get_bptr()
  auto node = OnodeTree::make_node(*this); 
  OnodeDelta delta;
  auto p = bl.cbegin();
  // all operands in delta are decoded from bl,
  // or blocks pointed by paddr(s) encoded in bl
  decode(delta, p);
  switch(delta.op) {
  case delta.op_t::insert_or_assign:
    return node.insert_at(delta.n, delta.oid, delta.onode);
  case delta.op_t::update_slot:
    return node.point_slot_to(delta.n, delta.oid);
  case delta.op_t:: shift_left:
    return node.shift_left(delta.n);
  case delta.op_t::truncate:
    return node.truncate(delta.n);
  case delta.op_t::insert_front:
    return node.push_right(delta.n, delta.keys, delta.items);
  case delta.op_t::insert_back:
    return node.append(delta.n, delta.keys, delta.items);
  case delta.op_t::remove:
    return node.insert_at(delta.n);
  default:
    assert(0 == "unknown onode delta");
  }
```

Reply:
Yeah, that's the notion.

Actually, base is completely irrelevant to LogicalCachedExtents -- you should ignore it.  Blocks that use physical addresses need it to resolve final addresses for contained relative addresses (that it, only LBA nodes and the root node itself).

Reply:
thanks, Sam!